pkg/automation/apply_test.go

```
@@ -0,0 +1,1026 @@
+	for i := range received {
+		assert.Equal(t, expected[i].functionName, received[i].functionName, "a wrong function was called")
+
+		assert.Equal(t, len(expected[i].arguments), len(received[i].arguments), fmt.Sprintf("function %s was called with a wrong number of arguments", expected[i].functionName))
+		for j := range received[i].arguments {
+			assert.Equal(t, expected[i].arguments[j], received[i].arguments[j], fmt.Sprintf("function %s was called with a wrong argument", expected[i].functionName))
+		}
+
+		assert.Equal(t, len(expected[i].results), len(received[i].results), fmt.Sprintf("function %s returned wrong number of values", expected[i].functionName))
+		for j := range received[i].results {
+			assert.Equal(t, expected[i].results[j], received[i].results[j], fmt.Sprintf("function %s returned a wrong result", expected[i].functionName))
+		}
+	}
+}
+
+// Checks if the test machine type operation works as expected.
+func TestTestMachineTypeOperation(t *testing.T) {
+	operation := gcloudOperation{
+		Action:       "test",
+		Path:         "/machineType",
+		Resource:     "//compute.googleapis.com/projects/rightsizer-test/zones/us-east1-b/instances/alicja-test",
+		ResourceType: "compute.googleapis.com/Instance",
+		ValueMatcher: &gcloudValueMatcher{MatchesPattern: ".*zones/us-east1-b/machineTypes/n1-standard-4"},
+	}
+
+	service := ApplyMockService{getInstanceResult: &compute.Instance{MachineType: "zones/us-east1-b/machineTypes/n1-standard-4"}}
+	err := DoOperation(&service, &operation)
+	assert.Nilf(t, err, "DoOperation shouldn't return an error")
+
+	expectedFunctions := []string{"GetInstance"}
+	expectedArguments := [][]interface{}{{"rightsizer-test", "us-east1-b", "alicja-test"}}
+	expectedResults := [][]interface{}{{&compute.Instance{MachineType: "zones/us-east1-b/machineTypes/n1-standard-4"}, nil}}
+
+	expected, _ := newCalledFunctions(expectedFunctions, expectedArguments, expectedResults)
+	compareCalledFunctions(t, expected, service.calledFunctions)
```

put this code in 'if' as in this example: https://godoc.org/github.com/stretchr/testify/assert#NoError
because we don't want to check it if checking for no errors already failed

---

Actually, I wanted to make the code simpler, by just assuming, that as this is a helper function, we will never receive an error. So then maybe I should just cause panic, if something is wrong, instead of returning an error? I mean, it will only return an error, if it receives arrays of different lengths.

---

in my opinion you suggestion is okay, just leave some clear description of error in log.Fatal()